src/coreclr/tools/Common/Internal/Runtime/EETypeBuilderHelpers.cs

```diff
@@ -108,9 +108,33 @@ public static ushort ComputeFlagsEx(TypeDesc type)
                 flagsEx |= (ushort)EETypeFlagsEx.HasEagerFinalizerFlag;
             }
 
+            if (HasCriticalFinalizer(type))
+            {
+                flagsEx |= (ushort)EETypeFlagsEx.HasCriticalFinalizerFlag;
+            }
+
             return flagsEx;
         }
 
+        private static bool HasCriticalFinalizer(TypeDesc type)
+        {
+            if (type is MetadataType mdType &&
+                            mdType.Module == mdType.Context.SystemModule &&
+                            mdType.Name == "CriticalFinalizerObject" &&
+                            mdType.Namespace == "System.Runtime.ConstrainedExecution")
+            {
+                return true;
+            }
+            else if (type.HasBaseType)
+            {
+                return HasCriticalFinalizer(type.BaseType);
```

Comment:
it would be nice to avoid recursion here. in abnormal cases there could be a lot of bases.

Reply:
```
private static bool HasCriticalFinalizer(TypeDesc type)
{
    do
    {
        if (!type.HasFinalizer)
            return false;

        if (type is MetadataType mdType ....
            return true;

        type = type.BaseType;
    }
    while (type != null);

    return false;
}
```